File: lessons/dsa/prefix-sums/examples/range-sums.test.ts
import { expect, test } from "vitest";

import {
  buildPrefix,
  rangeTotalByPrefix,
  rangeTotalByScan,
} from "./range-sums.js";

const VALUES = [3, 1, 4, 1, 5, 9];

test("both approaches agree on every range", () => {
  const prefix = buildPrefix(VALUES);
  const disagreements: string[] = [];

  for (let start = 0; start < VALUES.length; start += 1) {
    for (let end = start; end < VALUES.length; end += 1) {
      if (
        rangeTotalByPrefix(prefix, start, end) !==
        rangeTotalByScan(VALUES, start, end)
      ) {
        disagreements.push(`${start}..${end}`);
      }
    }
  }

  expect(disagreements).toEqual([]);
});

test("the lesson's range totals ten", () => {
  expect(rangeTotalByPrefix(buildPrefix(VALUES), 2, 4)).toBe(10);
});

test("a range starting at zero needs no special case", () => {
  expect(rangeTotalByPrefix(buildPrefix(VALUES), 0, 0)).toBe(3);
});

test("the whole sequence", () => {
  expect(rangeTotalByPrefix(buildPrefix(VALUES), 0, VALUES.length - 1)).toBe(
    23,
  );
});

test("prefix is one longer than the values", () => {
  expect(buildPrefix(VALUES)).toHaveLength(VALUES.length + 1);
});

test("an empty sequence has a single zero prefix", () => {
  expect(buildPrefix([])).toEqual([0]);
});

test("negative values still subtract correctly", () => {
  expect(rangeTotalByPrefix(buildPrefix([5, -3, 2]), 0, 2)).toBe(4);
});

File: lessons/dsa/prefix-sums/examples/range_sums_test.go
package examples

import "testing"

var values = []int{3, 1, 4, 1, 5, 9}

func TestBothApproachesAgreeOnEveryRange(t *testing.T) {
	prefix := BuildPrefix(values)
	for start := range values {
		for end := start; end < len(values); end++ {
			byPrefix := RangeTotalByPrefix(prefix, start, end)
			byScan := RangeTotalByScan(values, start, end)
			if byPrefix != byScan {
				t.Fatalf("range %d..%d: prefix gave %d, scan gave %d", start, end, byPrefix, byScan)
			}
		}
	}
}

func TestTheLessonRangeTotalsTen(t *testing.T) {
	if got := RangeTotalByPrefix(BuildPrefix(values), 2, 4); got != 10 {
		t.Fatalf("expected 10, got %d", got)
	}
}

func TestARangeStartingAtZeroNeedsNoSpecialCase(t *testing.T) {
	if got := RangeTotalByPrefix(BuildPrefix(values), 0, 0); got != 3 {
		t.Fatalf("expected 3, got %d", got)
	}
}

func TestTheWholeSequence(t *testing.T) {
	if got := RangeTotalByPrefix(BuildPrefix(values), 0, len(values)-1); got != 23 {
		t.Fatalf("expected 23, got %d", got)
	}
}

func TestPrefixIsOneLongerThanTheValues(t *testing.T) {
	if got := len(BuildPrefix(values)); got != len(values)+1 {
		t.Fatalf("expected %d entries, got %d", len(values)+1, got)
	}
}

func TestAnEmptySequenceHasASingleZeroPrefix(t *testing.T) {
	prefix := BuildPrefix(nil)
	if len(prefix) != 1 || prefix[0] != 0 {
		t.Fatalf("expected a single zero, got %v", prefix)
	}
}

func TestNegativeValuesStillSubtractCorrectly(t *testing.T) {
	if got := RangeTotalByPrefix(BuildPrefix([]int{5, -3, 2}), 0, 2); got != 4 {
		t.Fatalf("expected 4, got %d", got)
	}
}
